Add ability to start live activities

nakajima opened the pull request:
I believe this PR brings things in line with the JSON payload described in [the docs](https://developer.apple.com/documentation/activitykit/starting-and-updating-live-activities-with-activitykit-push-notifications#Construct-the-payload-that-starts-a-Live-Activity).

This makes the `APNSLiveActivityNotificationEvent` type a bit more clunky to deal with, so I'd love to hear if anyone has any ideas on how to improve that.

I also ran `swift-format --configuration .swift-format` on my changes which seems to have brought in some unrelated format changes. If this is a PR worth pursuing I'm happy to clean those up.

See also https://github.com/swift-server-community/APNSwift/pull/191.

nakajima commented:
Ok actually it looks like I need to include an `alert` as well. On it.

nakajima commented:
OK, I've verified that this is working in a test app so it's ready for review!

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSLiveActivityNotificationEvent.swift:
```diff
@@ -12,15 +12,65 @@
-    
-    /// Specifies that live activity should be ended
-    public static let end = Self(rawValue: "end")
+public protocol APNSLiveActivityNotificationEvent: Hashable, Encodable {
```
This is a breaking change and we can't just change from a struct to a protocol. We should try and continue to model it with the current struct type.

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSLiveActivityNotification.swift:
```diff
@@ -17,7 +17,9 @@ import struct Foundation.UUID
 ///
 /// It is **important** that you do not encode anything with the key `aps`.
-public struct APNSLiveActivityNotification<ContentState: Encodable>: APNSMessage {
+public struct APNSLiveActivityNotification<ContentState: Encodable & Hashable & Sendable>:
```
We can't just add these requirements to the generic type. This is a breaking API change otherwise. We can make the conformance to those protocols conditional though and that should work.

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSLiveActivityNotification.swift:
```diff
@@ -30,34 +32,62 @@ public struct APNSLiveActivityNotification<ContentState: Encodable>: APNSMessage
+                return APNSLiveActivityNotificationEventEnd()
+            case "update":
+                return APNSLiveActivityNotificationEventUpdate()
+            default:
```
We should be exhaustive here and avoid using default cases if possible.

FranzBusch reviewed (commented):
Thanks for working on this. I left a few comments on the PR. Mostly this is a breaking API change so we have to do it slightly differently to avoid that breakage.

nakajima commented:
@FranzBusch Thanks so much for the review! I've moved things around to just introduce a separate object for the additional data a `start` event needs. I think the one downside to this approach is that if you pass a `start` event without the `startOptions`, the push just won't get sent. I don't think that's the end of the world though since that's also what APNS does afaik.

> We can't just add these requirements to the generic type. This is a breaking API change otherwise. We can make the conformance to those protocols conditional though and that should work.

Fixed in https://github.com/swift-server-community/APNSwift/pull/196/commits/5af6d92d57faddf3e9412303a8df81e383a87299 and https://github.com/swift-server-community/APNSwift/pull/196/commits/fffc9eaa7f9eaa95892c2e6c52080773e08f43d8.

> This is a breaking change and we can't just change from a struct to a protocol. We should try and continue to model it with the current struct type.

Fixed in https://github.com/swift-server-community/APNSwift/pull/196/commits/4671d0ada8a65122fd7ee33e8916d5ec5439bf15.

nakajima commented:
Actually this PR is wrong. I think we need to pass the ActivityAttributes as the `attributes` field, not the content state.

alex-taffe commented on Sources/APNSCore/LiveActivity/APNSLiveActivityNotification.swift:
```diff
@@ -152,7 +151,7 @@ public struct APNSLiveActivityNotification<ContentState: Encodable>: APNSMessage
         self.aps = APNSLiveActivityNotificationAPSStorage(
             timestamp: timestamp,
-            event: event.rawValue,
+						event: event.rawValue,
```
Just before this gets merged, funky indentation here

nakajima replied:
Fixed in https://github.com/swift-server-community/APNSwift/pull/196/commits/2d41546e2d7ffd6df8c2bba7bd3c25cf8ce7c3c2.

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSStartLiveActivityNotification.swift:
```diff
@@ -0,0 +1,176 @@
+/// A notification that starts a live activity
+///
+/// It is **important** that you do not encode anything with the key `aps`.
+public struct APNSStartLiveActivityNotification<Attributes: Encodable, ContentState: Encodable>:
+    APNSMessage
+{
```
This formatting looks weird. Can we fix this up?

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSStartLiveActivityNotificationAPSStorage.swift:
```diff
@@ -0,0 +1,51 @@
+//
+//===----------------------------------------------------------------------===//
+
+struct APNSStartLiveActivityNotificationAPSStorage<Attributes: Encodable, ContentState: Encodable>:
```
Same here

FranzBusch commented on Sources/APNSCore/LiveActivity/APNSStartLiveActivityNotification.swift:
```diff
@@ -0,0 +1,176 @@
+    ///   - attributes: The ActivityAttributes of the live activity to start
+    ///   - attributesString: The type name of the ActivityAttributes you want to send
+    ///   - alert: An alert that will be sent along with the notification
+    public init(
```
Do we need two inits?

nakajima replied:
No I don't think we do, removed in https://github.com/swift-server-community/APNSwift/pull/196/commits/a142abee3531d8297a28d69e3a7f492c75cc3d10.

FranzBusch reviewed (commented):
Looks good. Just one question

chrisschlitt commented:
Can confirm this works, was able to start a live activity from a push notification with:
```
let notification = APNSStartLiveActivityNotification(
    expiration: .immediately,
    priority: .immediately,
    appID: "com.example.example",
    contentState: TestContentState(),
    timestamp: 1_672_680_658,
    attributes: TestAttributes(),
    attributesType: "TestAttributes",
    alert: .init(title: .raw("Test"), body: .raw("Test"))
)
do {
    let response = try await apnsClient.send(
        APNSRequest(
            message: notification,
            deviceToken: kit.token,
            pushType: .liveactivity,
            expiration: notification.expiration,
            priority: notification.priority,
            apnsID: notification.apnsID,
            topic: notification.topic,
            collapseID: nil
        )
    )
    // Success
} catch {
    // Failure
}
```

alex-taffe commented:
I know this is somewhat off topic, but while everyone is here and I can solicit advice, [Apple's docs](https://developer.apple.com/documentation/activitykit/starting-and-updating-live-activities-with-activitykit-push-notifications#Start-new-Live-Activities-with-ActivityKit-push-notifications) state:
> When the system receives the ActivityKit push notification on a device, it starts a new Live Activity, wakes up your app, and grants it background run time to allow you to download assets that the Live Activity needs.

Does anyone know what method is invoked or what happens when the start notification goes out? I'm severely doubting it's just `application:didFinishLaunchingWithOptions:` but I could be wrong. I'm going to play around with it because I need it to download image assets, but just figured I'd just check to see if anyone has already figured it out
